include/myst/thread.h

```diff
@@ -96,18 +86,99 @@ struct myst_thread
     /* the process identifier (inherited from main thread) */
     pid_t pid;
 
-    /* unique thread identifier (same as pid for main thread) */
-    pid_t tid;
-
-    /* the value returned by gettid() on the target (for this thread) */
-    pid_t target_tid;
-
     /* The exit status passed to SYS_exit */
     int exit_status;
 
     /* Terminating signal value */
     unsigned terminating_signum;
 
+    /* the stack that was created by myst_exec() */
+    void* exec_stack;
+    size_t exec_stack_size;
+
+    /* the copy of the CRT data made by myst_exec() */
+    void* exec_crt_data;
+    size_t exec_crt_size;
+
+    /* lock when enumerating all threads in this process
+        while enumerating over thread->group_prev/next */
+    myst_spinlock_t thread_group_lock;
+
+    /* use this lock when using */
+    /* myst_process_list_lock */
+    myst_process_t* prev_process;
+    myst_process_t* next_process;
+
+    /* process CWD. Can be set on differnet threads so need to protect it too
+     */
+    char* cwd;
+    myst_spinlock_t cwd_lock;
+
+    /* The current umask this process */
+    mode_t umask;
+    myst_spinlock_t umask_lock;
+
+    /* the process group ID */
+    pid_t pgid;
+
+    /* Pointer to the main process thread */
+    myst_thread_t* main_process_thread;
+
+    /* the file-descriptor table is inherited from process thread */
+    myst_fdtable_t* fdtable;
+
+    /* doubly-linked zombie-list */
+    struct myst_process* zombie_next;
+    struct myst_process* zombie_prev;
+
+    /* If the clone is vfork mode, we may need to signal the parents
+     * initiation thread when an exec or exit is called from the child.
+     *
+     * This is necessary when pseudo fork is in wait_exec_exit mode. This
+     * also may be used for actual vfork().
+     */
+    bool is_pseudo_fork_process;
+    pid_t vfork_parent_pid;
+    pid_t vfork_parent_tid;
+
+    /* This is a copy of the uid from the main process thread and is only needed
+     * when dealing with zombie threads because the main thread is not available
+     */
+    uid_t process_uid;
+
+    struct
+    {
+        /* process-wide protection for manipulating signals that use either mask
+         * or sigactions*/
+        myst_spinlock_t lock;
+
+        // the signal handles registered through sigaction and
+        // shared by threads in the prcoess.
+        posix_sigaction_t* sigactions;
+
+        /* The mask of blocked signals */
+        uint64_t mask;
```

Comment:
Acc to the man page, it looks the mask is per-thread.
https://man7.org/linux/man-pages/man2/sigprocmask.2.html

Reply:
yeah! I discovered that the hard way! Fixed